### PR TITLE
Write path for download permission graphs

### PR DESCRIPTION
**Pull request by noahmoss**
This PR contains the basic write path for download permissions.

This is pretty similar to the write path for data access permissions. However, there are a few key differences
* It lives in a new enterprise permissions namespace, ` metabase-enterprise.advanced-permissions.models.permissions`
* Values in the permissions graph for download perms can be `:full`, `:limited` or `:none`
* Native query download perms are fully inferred from non-native download perms. This is done by reading the list of tables in the DB and checking the download permission level for each one.
  * This is necessary because we don't have a separate toggle in the UI for native download perms. Rather, we automatically set it equal to the most permissive setting supported by _all_ tables.
  * This also means that we ignore any values in the permission graph passed by the frontend for native query download perms.
  
Things that are on my radar, that this PR doesn't address (these will come later):
* Updating native query download perms after sync, in case tables or schemas were added or removed
* A migration to add full download perms for existing groups & DBs
* Making sure that DBs have download perms when upgrading from OSS to EE

UPDATE 3/7: I've moved some helper functions, particularly `update-native-download-permissions!` and the functions that it depends on, into the OSS namespace. This is because I'll need to call them during sync (in the next PR) to keep native download perms consistent with non-native download perms, even on OSS instances.

Epic #20380

### CODE REVIEW

File: enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions_test.clj [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  (:require [clojure.test :refer :all]
+            [metabase-enterprise.advanced-permissions.models.permissions :as ee-perms]
+            [metabase.models :refer [Permissions PermissionsGroup]]
+            [metabase.models.database :as database]
+            [metabase.models.permissions :as perms]
+            [metabase.public-settings.premium-features-test :as premium-features-test]
+            [metabase.test :as mt]
+            [metabase.util :as u]))
+
+(defn- test-download-perms [group-id]
```
Comment (qnkhuat) [inline enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions_test.clj:11]: the name is somewhat confused to me, how about we call it `download-perms-by-group-id`?

File: src/metabase/models/permissions.clj [hunk 2]
```diff
@@ -776,7 +829,65 @@
...
-(s/defn ^:private update-table-read-perms!
+(defn- download-permissions-set
+  [group-id]
+  (map :object
+       (db/select [Permissions :object]
+                  {:where [:and
+                           [:= :group_id group-id]
+                           [:or
+                            [:= :object (hx/literal "/")]
+                            [:like :object (hx/literal "/download/%")]]]})))
```
Comment (qnkhuat) [inline src/metabase/models/permissions.clj:840]: what you have here is a list, not a set.

Or you can use `db/select-field`, it returns a set by default.
```suggestion
  (db/select-field :object [Permissions :object]
                   {:where [:and
                            [:= :group_id group-id]
                            [:or
                             [:= :object (hx/literal "/")]
                             [:like :object (hx/literal "/download/%")]]]}))
```

File: src/metabase/models/permissions.clj [hunk 3]
```diff
@@ -776,7 +829,65 @@
...
+   :full
+
+   (set-has-full-permissions? permissions-set (feature-perms-path :download :limited db-id schema-name table-id))
+   :limited
+
+   :else
+   :none))
+
+(s/defn update-native-download-permissions!
+  "To update native download permissions, we must read the list of tables in the database, and check the group's
```
Comment (qnkhuat) [inline src/metabase/models/permissions.clj:855]: I'd add a comment explaining: the native download permission will be applied to the entire database, not a single table.

File: enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions_test.clj (hunk 1 above)
Comment (noahmoss) [reply enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions_test.clj:11]: yeah, good call

File: src/metabase/models/permissions.clj (hunk 2 above)
Comment (noahmoss) [reply src/metabase/models/permissions.clj:840]: Good point, the code works the same for any kind of collection, but I like your revised query

**Review by camsaul — APPROVED** [review]

**Pull request merged**